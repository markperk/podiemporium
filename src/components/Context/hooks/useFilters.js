import { useBrix } from 'react-brix'
import { fromJS } from 'immutable'
import { paths } from '../paths'

export const useFilters = () => {
  const [genre, setGenre] = useBrix(paths.filters.genre.get())
  const [playId, setPlayId] = useBrix(paths.filters.playId.get())
  const [rating, setRating] = useBrix(paths.filters.rating.get())
  const [year, setYear] = useBrix(paths.filters.year.get())
  const range = rating && [rating.get(0), rating.get(1)]

  let filters = {}
  if (genre) filters.genre = [genre]
  if (playId) filters.play_id = [playId]
  if (year) filters.year = [year]
  if (rating) filters.rating = range

  return [
    fromJS(filters),
    {
      useGenre: [genre, setGenre],
      usePlayId: [playId, setPlayId],
      useRating: [range, setRating],
      useYear: [year, setYear],
    }
  ]
}