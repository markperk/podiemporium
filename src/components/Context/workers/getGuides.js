import { tutorialData } from '../../../utils/tutorialData'


export const getGuides = async () => {
  return await tutorialData;
}