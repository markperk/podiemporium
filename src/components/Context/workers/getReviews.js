import { api } from '../actions'

const assignPlayId = reviews => reviews.map(review => {
  review.play_id = Math.floor(Math.random() * 37) + 1
  return review
})

export const getReviews = async () => {
  return await api.get('/reviews').then(reviews => assignPlayId(reviews))
}