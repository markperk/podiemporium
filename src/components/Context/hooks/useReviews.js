import { List } from 'immutable'
import { useBrixWorker } from 'react-brix'
import { paths } from '../paths'
import { getReviews } from '../workers'
import { useFilters } from '../hooks'
import { filterBy } from '../../../utils'

export const useReviews = () => {
  const [filters, filterSetters] = useFilters()

  const worker = async () => {
    try {
      return await getReviews()
    }
    catch (error) {
      console.error(error)
    }
  }

  const reviews = useBrixWorker(paths.reviews.get(), worker, List())
  const filteredReviews = filterBy(reviews, filters)
  const filterCount = [filteredReviews.size, reviews.size]

  return [
    filteredReviews,
    filterSetters,
    filterCount,
  ]
}