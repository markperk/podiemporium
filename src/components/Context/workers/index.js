import { getReviews } from './getReviews'

export {
  getReviews,
}