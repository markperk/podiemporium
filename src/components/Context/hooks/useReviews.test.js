import { fromJS } from 'immutable'
import { useBrixWorker } from 'react-brix'
import { useReviews } from './useReviews'
import { useFilters } from './useFilters'
import { getReviews } from '../workers/getReviews'

jest.mock('react')
jest.mock('react-brix')
jest.mock('../workers/getReviews')
jest.mock('./useFilters')

const mockReviews = [{
  author: "Kaley Schiller",
  body: "The fool doth think he is wise, but the wise man knows himself to be a fool.",
  id: "9783221620868",
  publish_date: "2016-09-05T23:25:47.642350Z",
  rating: 0.8,
  play_id: 1,
}]


describe('useReviews', () => {

  describe('filteredReviews && filterSetters', () => {
    let useRating, usePlayId, useGenre,
      filters, filterSetters

    beforeEach(() => {
      jest.clearAllMocks()
      filters = fromJS({
        genre: 'comedy',
      })
      useGenre = ['comedy', jest.fn()]
      usePlayId = ['1', jest.fn()]
      useRating = [[0.8, 2.0], jest.fn()]
      filterSetters = { useGenre, usePlayId, useRating }
      useFilters.mockReturnValue([filters, filterSetters])
      getReviews.mockReturnValue(mockReviews)
      useBrixWorker.mockReturnValue(fromJS(mockReviews))
    })

    test('returns an empty array of reviews', () => {
      const result = useReviews()
      expect(result).toMatchSnapshot()
    })
  })

  describe('filteredReviews match', () => {
    let useRating, usePlayId, useGenre,
      filters, filterSetters

    beforeEach(() => {
      jest.clearAllMocks()
      filters = fromJS({
        rating: [0.8, 2.0],
      })
      useGenre = ['comedy', jest.fn()]
      usePlayId = ['1', jest.fn()]
      useRating = [[0.8, 2.0], jest.fn()]
      filterSetters = { useGenre, usePlayId, useRating }
      useFilters.mockReturnValue([filters, filterSetters])
      getReviews.mockReturnValue(mockReviews)
      useBrixWorker.mockReturnValue(fromJS(mockReviews))
    })

    test('returns an array of a review and setter hooks', () => {
      const [filteredReviews] = useReviews()
      expect(filteredReviews).toEqual(fromJS(mockReviews))
    })
  })

  describe('filterCount', () => {
    let useRating, usePlayId, useGenre,
      filters, filterSetters

    beforeEach(() => {
      jest.clearAllMocks()
      filters = fromJS({
        play_id: '1',
      })
      useGenre = ['comedy', jest.fn()]
      usePlayId = ['1', jest.fn()]
      useRating = [[0.8, 2.0], jest.fn()]
      filterSetters = { useGenre, usePlayId, useRating }
      useFilters.mockReturnValue([filters, filterSetters])
      getReviews.mockReturnValue(mockReviews)
      useBrixWorker.mockReturnValue(fromJS(mockReviews))
    })

    test('returns reviews, setter fns and a filterCount', () => {
      const [filteredReviews, filterSetters, filterCount] = useReviews()
      expect(filteredReviews).toEqual(fromJS(mockReviews))
      expect(filterSetters).toMatchSnapshot()
      expect(filterCount).toEqual([1, 1])
    })
  })
})