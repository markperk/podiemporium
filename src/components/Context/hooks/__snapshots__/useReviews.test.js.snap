// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useReviews filterCount returns reviews, setter fns and a filterCount 1`] = `
Object {
  "useGenre": Array [
    "comedy",
    [MockFunction],
  ],
  "usePlayId": Array [
    "1",
    [MockFunction],
  ],
  "useRating": Array [
    Array [
      0.8,
      2,
    ],
    [MockFunction],
  ],
}
`;

exports[`useReviews filteredReviews && filterSetters returns an empty array of reviews 1`] = `
Array [
  Immutable.List [],
  Object {
    "useGenre": Array [
      "comedy",
      [MockFunction],
    ],
    "usePlayId": Array [
      "1",
      [MockFunction],
    ],
    "useRating": Array [
      Array [
        0.8,
        2,
      ],
      [MockFunction],
    ],
  },
  Array [
    0,
    1,
  ],
]
`;
